File: problems/Rotate Image/rotate.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    const len = matrix.length - 1;

    for (let i = 0; 1 + (len >> 1) > i; ++i) {
        for (let j = i; len - i > j; ++j) {
            let tmp = matrix[i][j];
            matrix[i][j] = matrix[len-j][i];
            [matrix[len-j][i], matrix[len-i][len-j]] = [matrix[len-i][len-j], matrix[j][len-i]];
            matrix[j][len-i] = tmp;
        }
    }
};

File: problems/Rotate Image/rotate.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        for (int i = 0; (n + 1) / 2 > i; ++i) { // 0, 0
            for (int j = 0; n / 2 > j; ++j) { // 0, 1
                int tmp = matrix[n - 1 - j][i]; // 7, 4
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j]; // 9, 8
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i]; // 3, 6
                matrix[j][n - 1 - i] = matrix[i][j]; // 1, 2
                matrix[i][j] = tmp; // 7, 4
            }
        }
    }
}
